src/autocomplete/Autocompleter.ts

```diff
@@ -82,15 +82,22 @@ export default class Autocompleter {
         });
     }
 
-    async getCompletions(query: string, selection: ISelectionRange, force = false): Promise<IProviderCompletions[]> {
+    // eslint-disable-next-line
+    async getCompletions(query: string, selection: ISelectionRange, force = false, limit = -1): Promise<IProviderCompletions[]> {
         /* Note: This intentionally waits for all providers to return,
          otherwise, we run into a condition where new completions are displayed
          while the user is interacting with the list, which makes it difficult
          to predict whether an action will actually do what is intended
         */
         // list of results from each provider, each being a list of completions or null if it times out
-        const completionsList: ICompletion[][] = await Promise.all(this.providers.map(provider => {
-            return timeout(provider.getCompletions(query, selection, force), null, PROVIDER_COMPLETION_TIMEOUT);
+        const completionsList: ICompletion[][] = await Promise.all(this.providers.map(async provider => {
+            const completions = await timeout(
+                provider.getCompletions(query, selection, force),
+                null,
+                PROVIDER_COMPLETION_TIMEOUT,
+            );
+            const maxCompletionsItems = limit > -1 ? limit : completions.length;
+            return completions.slice(0, maxCompletionsItems);
```

Comment:
might be sane to pass the limit through to the `getCompletions` here and have it do the slice before it instantiates a bunch of objects & components

Reply:
I did not do that because the command is sent to the `QueryMatcher` who will then find all matches and sort them by relevance.

The whether we limit or not, if we want to get the same quality of results I don't really see a way around creating those objects and then discarding the ones we don't want to render in the DOM

Reply:
I'm saying the `slice` call would be better placed before this map
https://github.com/matrix-org/matrix-react-sdk/blob/develop/src/autocomplete/UserProvider.tsx#L121
(in each provider)

This way we only create 20 `<PillCompletion`s and 20 `<MemberAvatar`s and not an unbounded amount only to discard all but 20

Reply:
For de-dup the `QueryMatcher::match` call maybe could take the optional limit param to not duplicate logic between all providers